doc/manual/src/language/string-context.md

```diff
@@ -0,0 +1,145 @@
+# String context
+
+> **Note**
+>
+> This is an advanced topic.
+> The Nix language is designed to be used without the programmer knowing string contexts are.
+
+Strings in the Nix language are not just a sequences of characters like strings in other languages.
+They are actually pairs of sequences of characters and what is known as a *string context*.
+The string context is an (unordered) set of *string context elements*.
+
+The goal of string contexts are to collect non-string values combined with strings via
+[string concatenation](./operators.md#string-concatenation),
+[string interpolation](./string-interpolation.md),
+and similar operations.
+The idea is that a user can combine together values to create a build recipe without manually keeping track of where the "ingredients" come from, and then the Nix language does that bookkeeping implicitly to come up with the right derivation inputs.
```

Comment:
I say "values" because arguably the things we are concatentaing aren't really strings proper. The string context in purpose a way to concatenate strings and *non strings* together --- by expanding the definition of "string" to accommodate these other things.

Reply:
```suggestion
The idea is that a user can combine values into build instructions for derivations without manually keeping track of where they come from.
Then the Nix language implicitly does that bookkeeping to efficiently obtain the closure of derivation inputs.
```